sockeye/inference.py

```diff
@@ -514,30 +552,37 @@ def get_max_output_length(input_length: int):
 
 
 Tokens = List[str]
-TranslatorInput = NamedTuple('TranslatorInput', [
-    ('id', int),
-    ('sentence', str),
-    ('tokens', Tokens),
-])
-"""
-Required input for Translator.
-
-:param id: Sentence id.
-:param sentence: Input sentence.
-:param tokens: List of input tokens.
-"""
 
-InputChunk = NamedTuple("InputChunk",
-                        [("id", int),
-                         ("chunk_id", int),
-                         ("tokens", Tokens)])
-"""
-A chunk of a TranslatorInput.
+class TranslatorInput:
+    """
+    Required input for Translator.
 
-:param id: Sentence id.
-:param chunk_id: The id of the chunk.
-:param tokens: List of input tokens.
-"""
+    :param id: Sentence id.
+    :param sentence: Input sentence.
+    :param tokens: List of input tokens.
+    :param factors: List of zero or more input factors.
+    """
+    def __init__(self, id, sentence, tokens, chunk_id = 0, factors = []):
+        self.id = id
+        self.chunk_id = chunk_id
+        self.sentence = sentence
+        self.tokens = tokens
+        self.factors = factors
+        self.num_factors = len(factors)
+
+    def __str__(self):
+        return '[%d.%d] %s' % (self.id, self.chunk_id, self.sentence)
+
+    def chunks(self, chunk_size: int):
+        """
+        Takes a TranslatorInput (itself) and splits it into chunks, in the case where the
+        input sentence is greater than the requested chunk_size.
+        """
+        for chunk_id, i in enumerate(range(0, len(self.tokens), chunk_size)):
+            new_tokens = self.tokens[i:i + chunk_size]
+            new_sent = ' '.join(new_tokens)
+            new_factors = [self.factors[f][i: i + chunk_size] for f in range(self.num_factors)]
+            yield TranslatorInput(id=self.id, chunk_id=chunk_id, sentence=new_sent, tokens=new_tokens, factors=new_factors)
```

Comment:
Since we have to carry factors around now, splitting for sentences that are too long is a bit more complicated, so this is now a class instead of a `namedtuple`. I also got rid of `InputChunk`, rolling it into `TranslatorInput`, which has an optional additional field, `chunk_id`.